### PR TITLE
prune outside states on failure inside container

### LINKED ISSUE #247: Pruning across container boundaries
In #221 we implemented correct pruning of infeasible candidate solutions inside SerialContainers.

However, this does not work across container boundaries yet because Containers currently act as opaque stages in their parent.

Consider this example structure:

rect2643

Assuming InterfaceState `a` fails to extend in Stage `Back`, `Fwd2` does not have to compute  InterfaceState `b` and the state can be disabled.
The same is true for the inverted case.

Parallel containers can result in similar situations but have to consider alternative paths through the different children as well.

I believe this is what you meant in https://github.com/ros-planning/moveit_task_constructor/pull/221#issuecomment-804847964 @rhaschke ?

### PR DESCRIPTION
**Pull request by v4hn**
Followup to #250 and should fix #247 (up to the additional pruning you mentioned in the last post @rhaschke).

### CODE REVIEW

**Review by rhaschke — COMMENTED** [review]
Very cool! Some minor remarks only.

File: core/src/container.cpp [hunk 1]
```diff
@@ -183,10 +183,8 @@ void ContainerBasePrivate::setStatus(const InterfaceState* s, InterfaceState::St
 	for (const SolutionBase* successor : trajectories<dir>(*s))
 		setStatus<dir>(state<dir>(*successor), status);
 }
-template void ContainerBasePrivate::setStatus<Interface::FORWARD>(const InterfaceState* s,
-                                                                  InterfaceState::Status status);
-template void ContainerBasePrivate::setStatus<Interface::BACKWARD>(const InterfaceState* s,
-                                                                   InterfaceState::Status status);
+template void ContainerBasePrivate::setStatus<Interface::FORWARD>(const InterfaceState*, InterfaceState::Status);
+template void ContainerBasePrivate::setStatus<Interface::BACKWARD>(const InterfaceState*, InterfaceState::Status);
```
Comment (v4hn) [inline core/src/container.cpp:187]: Some part of my toolchain keeps changing these lines back, sorry.
I just tried to find the offender though and everything behaves well, now that I look for the source... :confused:

File: core/include/moveit/task_constructor/container.h [hunk 2]
```diff
@@ -77,8 +77,6 @@ class ContainerBase : public Stage
 
 	/// called by a (direct) child when a new solution becomes available
 	virtual void onNewSolution(const SolutionBase& s) = 0;
-	/// called by a (direct) child when a solution failed
-	virtual void onNewFailure(const Stage& child, const InterfaceState* from, const InterfaceState* to) = 0;
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply core/include/moveit/task_constructor/container.h:81]: Now, an inherited container class cannot even override the behavior...
However, I agree that, currently, I don't see another use for this. However, in the future, we might see such use. 
And calling the parent's implementation of a certain functionality is common practice, isn't it?

[the remark this answers is not in the document]
Comment (v4hn) [reply core/include/moveit/task_constructor/container.h:81]: > Now, an inherited container class cannot even override the behavior...

That's the point of removing an unnecessary interface :smirk: .
Are you ok with making moving it back if we ever need it?

> And calling the parent's implementation of a certain functionality is common practice, isn't it?

Yes it is. But it's also common practice to wonder why some Qt-GUI slightly misbehaves after the developer forgot to call the parent. :smile:

[the remark this answers is not in the document]
Comment (rhaschke) [reply core/include/moveit/task_constructor/container.h:81]: > Are you ok with making moving it back if we ever need it?

I was struggling with this idea quite a while, but no, not really. 
Placing the API into the private interface will suggest (also for our future selves) that it should be forbidden to override the behavior of a container for `onNewFailure`. However, I cannot see why this should be the case.
If you don't have the time to move the API back into ContainerBase, I can do it. Just tell me.

**Review by rhaschke — APPROVED** [review]

File: core/test/test_serial.cpp [hunk 3]
```diff
@@ -138,234 +138,184 @@ unsigned int ForwardMockup::id_ = 0;
...
+		Connect::id_ = 0;
+	}
+	template <typename C, typename S>
+	auto add(C& container, S* stage) -> S* {
+		container.add(Stage::pointer(stage));
+		return stage;
+	}
+};
 
+using ConnectConnect = TestBase;
```
Comment (v4hn) [inline core/test/test_serial.cpp:162]: nifty!

**Pull request merged**